File: binarySearch7/FindMinimumInRotatedSortedArray.java

package binarySearch7;


// time complexity O(log(n))
// space complexity O(1)
public class FindMinimumInRotatedSortedArray {
    public int findMin(int[] nums) {
         int low = 0, high = nums.length - 1;
        int ans = Integer.MAX_VALUE;
      
        while (low <= high) {
            int mid = (low + high) / 2;

            //if left part is sorted:
            if (nums[low] <= nums[mid]) {
                // keep the minimum:
                ans = Math.min(ans, nums[low]);

                // Eliminate left half:
               
               low = mid+1;

            } else { //if right part is sorted:

                // keep the minimum:
                ans = Math.min(ans, nums [mid]);

                // Eliminate right half:
                high = mid - 1;
            }
        }
        return ans;
    }
    
    public static void main(String[] args) {
        FindMinimumInRotatedSortedArray obj = new FindMinimumInRotatedSortedArray();
        
         int [] nums = {3,4,5,1,2};
         System.out.println(obj.findMin(nums));
    }
}

File: LEETCODE128/Find Minimum in Rotated Sorted Array.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LEETCODE128
{


    internal class Find_Minimum_in_Rotated_Sorted_Array
    {
        //TC O(logn)
        // SC O(1)
        public int FindMin(int[] nums)
        {
            int left = 0; 
             int right = nums.Length - 1;
            int min = int.MaxValue;
             while (left <= right)
            {
                 int mid = left + (right - left)/2;
                if (nums[left] <= nums[mid])
                {
                    min = Math.Min(min, nums[left]);
                    left = mid + 1;
                }
                else
                {
                    min = Math.Min(min, nums[mid]);
                    right = mid - 1;
                }
            }
             return min;
        }
       /* static void Main(string[] args)
        {
            Find_Minimum_in_Rotated_Sorted_Array obj = new Find_Minimum_in_Rotated_Sorted_Array();
            int[] nums = { 3, 4, 5, 1, 2 };
            Console.WriteLine(obj.FindMin(nums));

        }*/

    }

    // Approach 1
    // TC O(n)
    //SC O(1)
    /* internal class Find_Minimum_in_Rotated_Sorted_Array
     {
         public int FindMin(int[] nums)
         {
             int min = int.MaxValue;
             for(int i = 0; i<nums.Length; i++)
             {
                 min = Math.Min(min, nums[i]);

             }
             return min;
         }
         static void Main(string[] args)
         {
             Find_Minimum_in_Rotated_Sorted_Array obj = new Find_Minimum_in_Rotated_Sorted_Array();
             int[] nums = {3, 4, 5, 1, 2};
             Console.WriteLine(obj.FindMin(nums));

         }

     }*/
}
